File: MTG Deck Editor/src/editor/gui/filter/editor/TypeLineFilterPanel.java
package editor.gui.filter.editor;

import javax.swing.BoxLayout;
import javax.swing.JTextField;

import editor.filter.Filter;
import editor.filter.FilterFactory;
import editor.filter.leaf.FilterLeaf;
import editor.filter.leaf.TypeLineFilter;
import editor.gui.generic.ComboBoxPanel;
import editor.util.Containment;

/**
 * This class represents a panel that corresponds to a filter that
 * groups cards by their entire type lines.
 * 
 * @author Alec Roelke
 */
@SuppressWarnings("serial")
public class TypeLineFilterPanel extends FilterEditorPanel<TypeLineFilter>
{
	/**
	 * Combo box for editing set containment.
	 */
	private ComboBoxPanel<Containment> contain;
	/**
	 * Text field for editing the type line to match.
	 */
	private JTextField line;
	
	/**
	 * Create a new TypeLineFilterPanel.
	 */
	public TypeLineFilterPanel()
	{
		super();
		setLayout(new BoxLayout(this, BoxLayout.X_AXIS));
		
		contain = new ComboBoxPanel<Containment>(Containment.values());
		add(contain);
		
		line = new JTextField();
		add(line);
	}
	
	/**
	 * Create a new TypeLineFilterPanel, using the given TypeLineFilter
	 * to initialize its fields.
	 * 
	 * @param f filter to use for initialization
	 */
	public TypeLineFilterPanel(TypeLineFilter f)
	{
		this();
		setContents(f);
	}
	
	@Override
	public Filter filter()
	{
		TypeLineFilter filter = (TypeLineFilter)FilterFactory.createFilter(FilterFactory.TYPE_LINE);
		filter.contain = contain.getSelectedItem();
		filter.line = line.getText();
		return filter;
	}

	@Override
	public void setContents(FilterLeaf<?> filter) throws IllegalArgumentException
	{
		if (filter instanceof TypeLineFilter)
			setContents((TypeLineFilter)filter);
		else
			throw new IllegalArgumentException("Illegal type line filter " + filter.type());
	}

	@Override
	public void setContents(TypeLineFilter filter)
	{
		contain.setSelectedItem(filter.contain);
		line.setText(filter.line);
	}
}

File: src/main/scala/editor/gui/filter/editor/TypeLineFilterPanel.scala
package editor.gui.filter.editor

import editor.database.attributes.CardAttribute
import editor.filter.leaf.FilterLeaf
import editor.filter.leaf.TypeLineFilter
import editor.gui.filter.FilterSelectorPanel
import editor.gui.generic.ComboBoxPanel
import editor.util.Containment

import javax.swing.BoxLayout
import javax.swing.JTextField

/**
 * Convenience constructors for [[TypeLineFilterPanel]].
 * @author Alec Roelke
 */
object TypeLineFilterPanel {
  /** @return a new [[TypeLineFilterPanel]] with no text */
  def apply(selector: FilterSelectorPanel) = new TypeLineFilterPanel(selector)

  /**
   * Create a new [[TypeLineFilterPanel]] with pre-populated text.
   * 
   * @param filter filter to get the text and containment from
   * @return a new [[TypeLineFilterPanel]] with the given filter's text and containment mode
   */
  def apply(filter: TypeLineFilter, selector: FilterSelectorPanel) = {
    val panel = new TypeLineFilterPanel(selector)
    panel.setContents(filter)
    panel
  }
}

/**
 * A filter editor panel for creating filters for all values that could appear on a card's type line.
 * @author Alec Roelke
 */
class TypeLineFilterPanel(selector: FilterSelectorPanel) extends FilterEditorPanel[TypeLineFilter] {
  setLayout(BoxLayout(this, BoxLayout.X_AXIS))

  private val contain = ComboBoxPanel(Containment.values)
  add(contain)

  private val line = JTextField()
  add(line)

  protected override val attribute = CardAttribute.TypeLine

  override def filter = TypeLineFilter(faces = selector.faces, contain = contain.getSelectedItem, line = line.getText)

  override def setFields(filter: TypeLineFilter) = {
    contain.setSelectedItem(filter.contain)
    line.setText(filter.line)
  }
}